Implement last baseline alignment for CSS Grid.

sammygill opened the pull request:
#### 7545fa38ae640521e6ab05ffc5cf73031c8aa521
<pre>
Implement last baseline alignment for CSS Grid.
<a href="https://bugs.webkit.org/show_bug.cgi?id=243524">https://bugs.webkit.org/show_bug.cgi?id=243524</a>
rdar://98089709

Reviewed by Brent Fulgham.

In order to implement last baseline alignment for grid, there are a
couple of extra pieces of information we need to keep track of:
1.) The grid items that are participating in last baseline alignment
2.) The descent values of each of those items

Once we have this, then the rest is similar to how first baseline
alignment is done.

When determining which items are participating in last baseline
alignment, we need to know if any of those items also span into other
tracks. This is because if an item spans into multiple tracks, then
it will participate in last baseline alignment with the items in the
last track it spans. Any calls to updateBaselineAlignmentContext must
now take this into consideration to provide the correct span to update
the baseline sharing groups.

Computing the descent values for last baseline items is the same idea
as computing the ascent for baseline items. This allows us to figure
out the &quot;shim&quot; value that is used to add an artificial margin to items
so that all of the last baseline items align. From the spec:

For the items in each baseline-sharing group, add a “shim”
(effectively, additional margin) on the start/end side
(for first/last-baseline alignment) of each item so that,
when start/end-aligned together their baselines align as specified.

Spec reference:
<a href="https://www.w3.org/TR/css-grid-1/#algo-baseline-shims">https://www.w3.org/TR/css-grid-1/#algo-baseline-shims</a>
<a href="https://www.w3.org/TR/css-grid-1/#grid-baselines">https://www.w3.org/TR/css-grid-1/#grid-baselines</a>
<a href="https://www.w3.org/TR/css-align-3/#baseline-sharing-group">https://www.w3.org/TR/css-align-3/#baseline-sharing-group</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-fieldset-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-fieldset-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-fieldset-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-flex-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-flex-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-flex-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-flex-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-grid-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-grid-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-grid-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-line-clamp-001.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-line-clamp-002.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-line-clamp-003.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-overflow-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-overflow-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-overflow-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-justify-baseline-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-justify-baseline-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-justify-baseline-004-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-001-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-002-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-grid-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-grid-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-grid-003-expected.txt:

* Source/WebCore/rendering/GridBaselineAlignment.cpp:
(WebCore::GridBaselineAlignment::logicalAscentForChild const):
We need to take into consideration if the item is being last baseline
aligned here. If it is, then we need to compute its descent value
instead.

(WebCore::GridBaselineAlignment::ascentForChild const):
Similarly, depending on the alignment we will need to call into the
child&apos;s firstLineBaseline or lastLineBaseline method to get the first
line&apos;s baseline value or the last line&apos;s baseline value, respectively.

(WebCore::GridBaselineAlignment::updateBaselineAlignmentContext):
A little bit of cleanup was done here. Instead of calling
ascentForChild and then checking if we need to compute the descent, we
can instead call logicalAscentForChild which does the same for us.
Also, instead of polluting the contextsMap with nullptrs, we can first
try to get the item from the map and decide whether we need to insert
it into the map of update the baseline sharing group instead.

(WebCore::GridBaselineAlignment::baselineOffsetForChild const):
* Source/WebCore/rendering/GridBaselineAlignment.h:

* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::updateBaselineAlignmentContext):
(WebCore::GridTrackSizingAlgorithm::baselineOffsetForChild const):
In both of the cases above, we need to check if the item is being first
or last baseline aligned. If it is being last baseline aligned, then it
needs to be in the baseline sharing group for the last track it spans.

* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::lastLineBaseline const):
Implementation to get the last line baseline for grid items. Definition
of the baseline is in the spec reference above.

(WebCore::RenderGrid::getBaselineChild const):
(WebCore::RenderGrid::columnAxisPositionForChild const):
(WebCore::RenderGrid::columnAxisOffsetForChild const):
A little bit of cleanup to separate the GridAxisEnd (used for last
baseline alignment) from the GridAxisCenter code. This required pulling
out some of the local variables so that they could be used in both
scopes. (startPosition + offsetFromStartPosition) should move the item
to the end of the axis and columnAxisOffsetForChild should provide the
baseline shim to align the item to the other last baselines. If an item
needs to be moved to GridAxisEnd but is not participating in last
baseline alignment, then the shim should be 0.

* Source/WebCore/rendering/RenderGrid.h:

Canonical link: <a href="https://commits.webkit.org/255455@main">https://commits.webkit.org/255455@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9242fdc3c28ac70bb8500bfcfa15abfb511e9685

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92559 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1778 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23146 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102307 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1776 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30145 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84994 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98472 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98221 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1195 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79058 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/28151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36555 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34337 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17918 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3779 "Found 1 new test failure: compositing/backing/backing-store-attachment-animating-outside-viewport.html (failure)") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38206 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40531 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1720 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40115 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37075 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1461,6 +1461,44 @@ std::optional<LayoutUnit> RenderGrid::firstLineBaseline() const
+    unsigned numColumns = m_grid.numTracks(ForColumns);
+    unsigned numRows = m_grid.numTracks(ForRows);
+
+    for (int column = numColumns - 1; column > -1; column--) {
```
Should we always be iterating through rows, until we find one that has items?

It looks like `firstLineBaseline` also doesn't do this though, so probably no need to fix this right now.

These two functions are really similar, is it possible to share code for them? The original has some 'FIXME' comments that would be nice to have applied to both.

sammygill replied:
I added RenderGrid::getBaselineChild to help out with this. Let me know what you think! We could also try to modify synthesizedBaselineFromBorderBox so that it gets the direction on its own instead of passing it in if we want to try to clean it up further.

mattwoodrow commented on Source/WebCore/rendering/GridBaselineAlignment.cpp:
```diff
@@ -123,31 +126,25 @@ void GridBaselineAlignment::updateBaselineAlignmentContext(ItemPosition preferen
-        auto* context = addResult.iterator->value.get();
-        context->updateSharedGroup(child, preference, ascent);
-    }
+    if (!contextSearch)
```
Super minor nit, would prefer the positive condition branch first, maybe `if (auto* contextSearch = ....)`

sammygill replied:
Changed!

mattwoodrow commented on Source/WebCore/rendering/GridBaselineAlignment.cpp:
```diff
@@ -56,16 +57,18 @@ LayoutUnit GridBaselineAlignment::marginUnderForChild(const RenderBox& child, Gr
-    LayoutUnit ascent = ascentForChild(child, baselineAxis);
-    return isDescentBaselineForChild(child, baselineAxis) ? descentForChild(child, ascent, baselineAxis) : ascent;
+    LayoutUnit ascent = ascentForChild(child, baselineAxis, position);
+    return (isDescentBaselineForChild(child, baselineAxis) || position == ItemPosition::LastBaseline) ? descentForChild(child, ascent, baselineAxis) : ascent;
```
I'm really struggling to get my head around the existing logic here, particularly the implementation of `isDescentBaselineForChild`. Why is it only considering the physical horizontal direction?

Are you sure that the logical OR of the two conditions is correct, not XOR?

I realise most of this isn't new to your changes, but if you understand it well enough to add comments explaining the physical/logical coord switches, then that'd be super helpful for me.

sammygill replied:
isDescentBaselineForChild is a little hard for me to read as well. If I'm understanding it correctly, I don't think it's considering only the physical horizontal direction and it might be that isHorizontalBaselineAxis is misleadingly named?

If the baseline alignment axis is the column axis, then I think isHorizontalBaselineAxis will still return true if the grid is in a vertical writing mode, right? I'm still not sure why this requires us to use the descent value for alignment.

mattwoodrow reviewed (approved):
Looks great to me!

Informal review only, since I'm not a WebKit review sorry.

I've added Javier Fernandez as a reviewer, since they wrote most of the existing code.

brentfulgham commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-fieldset-001-expected.txt:
```diff
@@ -14,6 +14,12 @@ FAIL .target > * 2 assert_equals:
 FAIL .target > * 3 assert_equals:
 <div data-offset-y="40"><span></span></div>
-offsetTop expected 40 but got 20
-PASS .target > * 4
```
PASS -> FAIL. Is this work tracked by one of the other last-baseline patches?

sammygill replied:
Keeping track of this issue here: https://bugs.webkit.org/show_bug.cgi?id=246413

brentfulgham commented on LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-fieldset-003-expected.txt:
```diff
@@ -14,6 +14,12 @@ FAIL .target > * 2 assert_equals:
-offsetLeft expected 45 but got 25
-PASS .target > * 4
+offsetLeft expected 45 but got 0
+FAIL .target > * 4 assert_equals:
```
Ditto: Do we expect this new failure?

sammygill replied:
Fieldset baselines were defined in a recent conversation but I don't think they are in the spec yet: https://github.com/w3c/csswg-drafts/issues/7656

AFAIK we don't really implement this behavior so the PASS was not occurring intentionally. I created a new bug to keep track of this issue: https://bugs.webkit.org/show_bug.cgi?id=246413

sammygill committed:
Implement last baseline alignment for CSS Grid.
https://bugs.webkit.org/show_bug.cgi?id=243524
rdar://98089709

Reviewed by Brent Fulgham.

In order to implement last baseline alignment for grid, there are a
couple of extra pieces of information we need to keep track of:
1.) The grid items that are participating in last baseline alignment
2.) The descent values of each of those items

Once we have this, then the rest is similar to how first baseline
alignment is done.

When determining which items are participating in last baseline
alignment, we need to know if any of those items also span into other
tracks. This is because if an item spans into multiple tracks, then
it will participate in last baseline alignment with the items in the
last track it spans. Any calls to updateBaselineAlignmentContext must
now take this into consideration to provide the correct span to update
the baseline sharing groups.

Computing the descent values for last baseline items is the same idea
as computing the ascent for baseline items. This allows us to figure
out the "shim" value that is used to add an artificial margin to items
so that all of the last baseline items align. From the spec:

For the items in each baseline-sharing group, add a “shim”
(effectively, additional margin) on the start/end side
(for first/last-baseline alignment) of each item so that,
when start/end-aligned together their baselines align as specified.

Spec reference:
https://www.w3.org/TR/css-grid-1/#algo-baseline-shims
https://www.w3.org/TR/css-grid-1/#grid-baselines
https://www.w3.org/TR/css-align-3/#baseline-sharing-group

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-fieldset-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-fieldset-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-fieldset-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-flex-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-flex-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-flex-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-flex-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-grid-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-grid-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-grid-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-line-clamp-001.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-line-clamp-002.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-line-clamp-003.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-overflow-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-overflow-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-overflow-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-justify-baseline-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-justify-baseline-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-justify-baseline-004-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-001-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-002-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-table-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-grid-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-grid-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-grid-003-expected.txt:

* Source/WebCore/rendering/GridBaselineAlignment.cpp:
(WebCore::GridBaselineAlignment::logicalAscentForChild const):
We need to take into consideration if the item is being last baseline
aligned here. If it is, then we need to compute its descent value
instead.

(WebCore::GridBaselineAlignment::ascentForChild const):
Similarly, depending on the alignment we will need to call into the
child's firstLineBaseline or lastLineBaseline method to get the first
line's baseline value or the last line's baseline value, respectively.

(WebCore::GridBaselineAlignment::updateBaselineAlignmentContext):
A little bit of cleanup was done here. Instead of calling
ascentForChild and then checking if we need to compute the descent, we
can instead call logicalAscentForChild which does the same for us.
Also, instead of polluting the contextsMap with nullptrs, we can first
try to get the item from the map and decide whether we need to insert
it into the map of update the baseline sharing group instead.

(WebCore::GridBaselineAlignment::baselineOffsetForChild const):
* Source/WebCore/rendering/GridBaselineAlignment.h:

* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::updateBaselineAlignmentContext):
(WebCore::GridTrackSizingAlgorithm::baselineOffsetForChild const):
In both of the cases above, we need to check if the item is being first
or last baseline aligned. If it is being last baseline aligned, then it
needs to be in the baseline sharing group for the last track it spans.

* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::lastLineBaseline const):
Implementation to get the last line baseline for grid items. Definition
of the baseline is in the spec reference above.

(WebCore::RenderGrid::getBaselineChild const):
(WebCore::RenderGrid::columnAxisPositionForChild const):
(WebCore::RenderGrid::columnAxisOffsetForChild const):
A little bit of cleanup to separate the GridAxisEnd (used for last
baseline alignment) from the GridAxisCenter code. This required pulling
out some of the local variables so that they could be used in both
scopes. (startPosition + offsetFromStartPosition) should move the item
to the end of the axis and columnAxisOffsetForChild should provide the
baseline shim to align the item to the other last baselines. If an item
needs to be moved to GridAxisEnd but is not participating in last
baseline alignment, then the shim should be 0.

* Source/WebCore/rendering/RenderGrid.h:

Canonical link: https://commits.webkit.org/255455@main
